chain/network/src/codec.rs

```diff
@@ -74,12 +74,64 @@ pub fn bytes_to_peer_message(bytes: &[u8]) -> Result<PeerMessage, std::io::Error
     PeerMessage::try_from_slice(bytes)
 }
 
+fn peer_id_type_field_len(enum_var: u8) -> Option<usize> {
+    // 1 byte for enum variant, then some number depending on the
+    // public key type
+    match enum_var {
+        0 => Some(1 + 32),
+        1 => Some(1 + 64),
+        _ => None,
+    }
+}
+
+pub fn is_forward_tx(bytes: &[u8]) -> Option<bool> {
+    let peer_message_variant = *bytes.get(0)?;
+
+    // PeerMessage::Routed variant == 13
+    if peer_message_variant != 13 {
+        return Some(false);
+    }
+
+    let target_field_variant = *bytes.get(1)?;
+    let target_field_len = if target_field_variant == 0 {
+        // PeerIdOrHash::PeerId
+        let peer_id_variant = *bytes.get(2)?;
+        peer_id_type_field_len(peer_id_variant)?
+    } else if target_field_variant == 1 {
+        // PeerIdOrHash::Hash is always 32 bytes
+        32
+    } else {
+        return None;
+    };
```

Comment:
This is error-prone. It implicitly relies on enum variants order, borsh serialization, and structures layout here. There are so many things that can go wrong if we change anything in any of the underlying structure that I am not sure if it is worth the trouble. (As to the proposal to extend Borsh, I wonder how that could look like, but that would be at least maintainable and reliable)

Can we decouple deserialization into a separate actor, so we can scale it, and don't block this actor on deserialization while doing its job?

Reply:
Yes, as I said in the PR description, this solution is a little hacky and very brittle. However, I don't think there is another solution at this time. Putting deserialization into another actor does not help unless we can have a different actor for different message types (which would then still require some way to know the message type before deserializing). The issue is not that the peer actor becomes blocked, it is that if we deserialize messages in the order they are received then we introduce a lot of latency on important messages (e.g. block approval) due to long runs of many transaction messages in a row.

That said, as a temporary solution, I think this is ok because it is unlikely the various things this code depends on (enum order, borsh details, etc) will change because any such change would be a protocol change. All nodes must agree on how to communicate with each other, so changing a message format is a protocol change, and I think that should happen pretty rarely. I still agree that this code is a nightmare to maintain so we should not keep it around in the long term, but as a bandaid fix for a dos vector, I think it does the job.

Reply:
@frol to partially address your concern, this function is covered in the test that is part of the CI.